File: colcon_ws/src/pkg_cpp_examples/src/simple_publisher.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <chrono>

using namespace std::chrono_literals;

class SimplePub: public rclcpp::Node
{
public:
    SimplePub(): Node("simple_publisher"), counter_{0}
    {
        pub_ = this->create_publisher<std_msgs::msg::String>("chatter", 10);
        timer_ = this->create_wall_timer(1s, std::bind(&SimplePub::timerCallback, this));
        RCLCPP_INFO(get_logger(), "Publishing at 1 Hz");
    }

    // class methods
    void timerCallback()
    {
        auto msg = std_msgs::msg::String();
        msg.data = "Hello ROS 2 - counter " + std::to_string(counter_ ++);
        pub_->publish(msg);
    }

private:
    unsigned int counter_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char* argv[])
{   
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimplePub>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;

}

File: colcon_ws/src/pkg_python_examples/pkg_python_examples/simple_publisher.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String

class SimplePublisher(Node):
    """
    This class implements the example to show how publisher works
    in the ros2.
    """
    def __init__(self):
        # this is class constructor
        super().__init__(node_name="simple_publisher")
        self.publisher_ = self.create_publisher(String, "chatter", 10)
        self.counter_ = 0
        self.frequency_ = 1.0
        self.get_logger().info(f"Publishing at rate of {self.frequency_}")

        # create the timer object
        self.timer_ = self.create_timer(timer_period_sec=self.frequency_, callback=self.timer_callback)
    
    def timer_callback(self):
        msg = String()
        msg.data = f"Hello ROS2, counter: {self.counter_}"
        self.publisher_.publish(msg)
        self.counter_ += 1

def main():
    # start the ros communication
    rclpy.init()
    # create an instance of the class
    simple_pub = SimplePublisher()
    # spin the node to run continuously
    rclpy.spin(simple_pub)
    # destroy the node
    simple_pub.destroy_publisher()
    # shutdown the ros communication
    rclpy.shutdown()

if __name__ == "__main__":
    main()

